src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -123,40 +166,91 @@ private async Task ProcessQueueAsync()
+                        // still need to react to it here by throwing away solution updates
+                        if (ranToCompletion)
+                        {
+                            lastMutatedSolution = mutatedSolution ?? lastMutatedSolution;
+                        }
+                        else
+                        {
+                            OnRequestServerShutdown($"An error occured processing a mutating request and the solution is in an invalid state. Check LSP client logs for any error information.");
+                            break;
+                        }
+                    }
+                    else
                     {
-                        lastMutatedSolution = mutatedSolution ?? lastMutatedSolution;
+                        // Non mutating are fire-and-forget because they are by definition readonly. Any errors
+                        // will be sent back to the client but we can still capture errors in queue processing
+                        // via NFW, though these errors don't put us into a bad state as far as the rest of the queue goes.
+                        _ = work.CallbackAsync(context, cancellationToken).ReportNonFatalErrorAsync();
                     }
                 }
-                else
-                {
-                    // Non mutating request get given the current solution state, but are otherwise fire-and-forget
-                    _ = work.CallbackAsync(context, cancellationToken);
-                }
+            }
+            catch (Exception e) when (FatalError.ReportWithoutCrash(e))
+            {
+                OnRequestServerShutdown($"Error occured processing queue: {e.Message}.");
+            }
+        }
+
+        private void OnRequestServerShutdown(string message)
+        {
+            RequestServerShutdown?.Invoke(this, new RequestShutdownEventArgs(message));
+
+            _cancelSource.Cancel();
+            DrainQueue();
+        }
+
+        /// <summary>
+        /// Cancels all requests in the queue and stops the queue from accepting any more requests. After this method
+        /// is called this queue is essentially useless.
+        /// </summary>
+        private void DrainQueue()
+        {
+            // Tell the queue not to accept any more items
+            _queue.Complete();
+
+            // Spin through the queue and pass in our cancelled token, so that the waiting tasks are cancelled.
+            // NOTE: This only really works because the first thing that CallbackAsync does is check for cancellation
+            // but generics make it annoying to store the TaskCompletionSource<TResult> on the QueueItem so this
+            // is the best we can do for now. Ideally we would manipulate the TaskCompletionSource directly here
+            // and just call SetCanceled
+            while (_queue.TryDequeue(out var item))
+            {
+                _ = item.CallbackAsync(default, new CancellationToken(true));
```

Comment:
If we don't call the callback, then the original request message task is never completed so StreamJsonRpc would be holding on to things. We need to complete the TaskCompletionSource that we created, and because its a generic type with no base class, the only way to do so is calling the callback.

And no, there is no Clear or Dispose method :P